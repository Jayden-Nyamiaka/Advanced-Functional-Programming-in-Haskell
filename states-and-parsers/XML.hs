----------------------------------------------------------------------
-- Simple XML parser.
----------------------------------------------------------------------

-- We call this module `Main` for testing purposes.
-- In a real-world scenario, it would be called `XML`.
module Main where

import System.Environment
import System.Exit
import System.IO
import Text.Parsec
import Text.Parsec.String

----------------------------------------------------------------------
-- Datatypes.
----------------------------------------------------------------------

type Tag = String

-- A character entity.
data Entity = LT_E | GT_E | AMP_E
   deriving (Show)

-- A single element of an XML document.
data Elem =
    TextE String     -- raw text
  | EntE Entity      -- entity
  | FormE Tag [Elem] -- tagged data
  deriving (Show)

----------------------------------------------------------------------
-- Simple parsers.
----------------------------------------------------------------------

-- Helper function for Parse Entity that 
-- parses the specific entity type.
parseEntityType :: Parser Entity
parseEntityType = 
  try (string "lt" >> return LT_E) 
  <|> try (string "gt" >> return GT_E)
  <|> (string "amp" >> return AMP_E)
  <?> "valid entity"

-- Parse an entity.
parseEntity :: Parser Entity
parseEntity = do
  char '&'
  ent <- parseEntityType
  char ';'
  return ent
  <?> "entity"

-- Parse some text not containing any tags or entities.
parseText :: Parser String
parseText = many1 (noneOf "<>&") <?> "text"

-- Parse a tag name, which is a nonempty sequence of letters,
-- all alphabetic characters or digits (no symbols), either
-- lower-case or upper-case.
parseTag :: Parser Tag
parseTag = many1 alphaNum <?> "tag"

----------------------------------------------------------------------
-- Form parsers.
----------------------------------------------------------------------

-- Parse a single tagged form.
parseTagged :: Parser (Tag, [Elem])
parseTagged = do
  char '<'
  tg <- parseTag
  char '>'
  contents <- many1 parseElem
  char '<'
  char '/'
  string tg
  char '>'
  return (tg, contents)
  <?> "tagged form"

-- Parse a single XML expression.
parseElem :: Parser Elem
parseElem = 
  (parseEntity >>= return . EntE)
  <|> (parseText >>= return . TextE)
  <|> try (parseTagged >>= \(tg, elems) -> 
    return (FormE tg elems))
  <?> "XML element"

-- Parse a series of elements from a string representing the entire contents of
-- a file.
parseElemsFromFile :: Parser [Elem]
parseElemsFromFile = do
  xs <- many parseElem
  eof
  return xs 
  <?> "file of XML elements"

----------------------------------------------------------------------
-- Pretty-printer.
----------------------------------------------------------------------

indent :: Int -> String
indent i = replicate i ' '

-- Pretty-print an XML element.
ppXML :: Int -> Elem -> String
ppXML i (TextE t)  = indent i ++ "Text[\n" ++ t ++ "\n" ++ indent i ++ "]\n"
ppXML i (EntE e)   = indent i ++ "Entity[" ++ show e ++ "]"
ppXML i (FormE t es) = 
  indent i ++ "Form: {" ++ t ++ "}[\n" 
  ++ concatMap (\s -> ppXML (i + 2) s ++ "\n") es
  ++ indent i ++ "]{/" ++ t ++ "}\n"

-- Parse all expressions in a file and run the pretty-printer on them.
runPpXML :: FilePath -> IO ()
runPpXML f = do
  p <- parseFromFile parseElemsFromFile f
  case p of
    Left err -> putStrLn $ "ERROR: " ++ show err
    Right ss -> 
      mapM_ (\s -> do
        putStrLn (ppXML 0 s)
        putStrLn "") ss

----------------------------------------------------------------------
-- Tests.
----------------------------------------------------------------------

usage :: String -> IO ()
usage s = do
  hPutStrLn stderr $ "usage: " ++ s ++ " filename"

main :: IO ()
main = do
  progName <- getProgName
  args <- getArgs
  case args of
    [filename] -> runPpXML filename >> exitSuccess
    _ -> usage progName >> exitFailure
