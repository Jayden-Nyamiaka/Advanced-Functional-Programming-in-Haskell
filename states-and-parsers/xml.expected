Form: {h1}[
  Text[
XML tests
  ]

]{/h1}


Text[



]


Form: {p}[
  Text[
This is a test 
of a simple XML parser.
  ]

]{/p}


Text[



]


Form: {p}[
  Text[
Hamlet's soliloquy:
  ]

]{/p}


Text[



]


Form: {blockquote}[
  Text[


To 
  ]

  Form: {u}[
    Text[
be
    ]

  ]{/u}

  Text[
, or 
  ]

  Form: {b}[
    Text[
not
    ]

  ]{/b}

  Text[
 to be, 
  ]

  Form: {em}[
    Text[
THAT
    ]

  ]{/em}

  Text[
 is the question.  "Haste makes

  ]

  Form: {em}[
    Text[
waste
    ]

  ]{/em}

  Text[
", 
  ]

  Entity[AMP_E]
  Text[
 as my Mom always said -- and that's the 
  ]

  Entity[LT_E]
  Text[
truth
  ]

  Entity[GT_E]
  Text[
!


  ]

]{/blockquote}


Text[



]


Form: {p}[
  Text[
Forms can be 
  ]

  Form: {a}[
    Form: {b}[
      Form: {c}[
        Form: {d}[
          Form: {e}[
            Text[
nested
            ]

          ]{/e}

        ]{/d}

      ]{/c}

    ]{/b}

  ]{/a}

  Text[
 arbitrarily deeply as
long as the end tags match the start tags.
  ]

]{/p}


Text[




]


Form: {h2}[
  Text[
Writing and testing your code
  ]

]{/h2}


Text[



]


Form: {p}[
  Text[
All of your Haskell functions, even internal ones (functions defined inside
other functions), should have explicit type signatures.
  ]

]{/p}


Form: {p}[
  Text[
All of your code
should be tested using 
  ]

  Form: {code}[
    Text[
ghci
    ]

  ]{/code}

  Text[
 with the 
  ]

  Form: {code}[
    Text[
-W
    ]

  ]{/code}

  Text[
 (warnings
enabled) command-line option.  Any warnings will be considered errors, so make
sure your code does not give any warnings.
  ]

]{/p}


Form: {p}[
  Text[
We are supplying automated
test scripts for part A and part C.  The test script for part A is located 
  ]

  Form: {a}[
    Text[

here
    ]

  ]{/a}

  Text[
 and the one for part C is located 
  ]

  Form: {a}[
    Text[
here
    ]

  ]{/a}

  Text[
. You can run them from

  ]

  Form: {code}[
    Text[
ghci
    ]

  ]{/code}

  Text[
 by putting them in the same directory as your code (which
must have the correct module names 
  ]

  Form: {em}[
    Text[
i.e.
    ]

  ]{/em}

  Text[
 
  ]

  Form: {code}[
    Text[
lab2ab.hs
    ]

  ]{/code}

  Text[
 for
part A and 
  ]

  Form: {code}[
    Text[
SparseMatrix.hs
    ]

  ]{/code}

  Text[
 for part D).  Then do this:
  ]

]{/p}


Text[



]


Form: {pre}[
  Text[


  ]

  Form: {code}[
    Text[
Prelude
    ]

    Entity[GT_E]
    Text[
 :l ./SignedNatTests.hs
*SignedNatTests
    ]

    Entity[GT_E]
    Text[
 main
[output is printed...]
Prelude
    ]

    Entity[GT_E]
    Text[
 :l ./SparseMatrixTests.hs
*SparseMatrixTests
    ]

    Entity[GT_E]
    Text[
 main
[output is printed...]

    ]

  ]{/code}

  Text[


  ]

]{/pre}


Text[



]


Form: {p}[
  Text[
and then the tests will be run and the output will be printed.  You will
probably find it difficult to understand the test code (it relies on a very
cool group of modules collectively called 
  ]

  Form: {code}[
    Text[
QuickCheck
    ]

  ]{/code}

  Text[
 which
implement randomized testing, and another group of modules called

  ]

  Form: {code}[
    Text[
HUnit
    ]

  ]{/code}

  Text[
 which implement unit testing) but the comments in the test
script and the test script output after running the tests will let you know if
your tests should fail at any point.  The 
  ]

  Form: {code}[
    Text[
QuickCheck
    ]

  ]{/code}

  Text[
 modules
define a custom monad called 
  ]

  Form: {code}[
    Text[
Gen
    ]

  ]{/code}

  Text[
 to generate random examples of
data structures, which is an application of monads we haven't discussed in
class yet.  The test script cannot test for everything, but it should catch the
most serious problems.
  ]

]{/p}


Form: {p}[
  Text[
One trick you might want to try in case you have a
test failure which is difficult to debug is to use the 
  ]

  Form: {code}[
    Text[
trace
    ]

  ]{/code}

  Text[

function from the 
  ]

  Form: {code}[
    Text[
Debug.Trace
    ]

  ]{/code}

  Text[
 module.  This function takes a
string and a value, prints the string and returns the value.  The trick is that
it does this without being in the 
  ]

  Form: {code}[
    Text[
IO
    ]

  ]{/code}

  Text[
 monad, so it uses the dreaded

  ]

  Form: {code}[
    Text[
unsafePerformIO
    ]

  ]{/code}

  Text[
 function.  This is OK since all it does is print
output before returning a value.  For a particular value in a computation, you
can do this:
  ]

]{/p}


Text[



]


Form: {pre}[
  Text[


  ]

  Form: {code}[
    Text[

-- ... somewhere inside an expression, you want to print the value of x ...  
-- instead of just (x), write: (trace ("x = " ++ show x) x) 

    ]

  ]{/code}

  Text[
 

  ]

]{/pre}


Text[




]


